### PR TITLE
CWL Input Bindings into bash command

### PR DESCRIPTION
**Pull request by danbills**
This PR takes inputs marked w/ "inputBinding" positions and appends them to the bash script in the specified position.

### CODE REVIEW

File: cwl/src/main/scala/cwl/CwlExpressionCommandPart.scala [hunk 1]
```diff
@@ -43,3 +44,29 @@ case class CwlArgumentCommandPart(argument: CommandLineBinding) extends CommandP
...
+                           functions: IoFunctionSet,
+                           valueMapper: (WomValue) => WomValue) = {
+
+    val localInputsMap = inputsMap.map{
+      case (LocalName(localName), WomSingleFile(path)) => localName -> WomString(path)
+      case (LocalName(localName), value) => localName -> value
+    }
+
+
+    val womValue: WomValue = argument match {
```
Comment (cjllanwarne) [inline cwl/src/main/scala/cwl/CwlExpressionCommandPart.scala]: At first I thought this was a straight copy/paste from above, but it does seem to be fundamentally different. 

Since it's so hard to parse the chains of `Some`s and `None`s and `_`s, could you add comments on what situation each case is trying to hit?

File: cwl/src/main/scala/cwl/CwlExpressionCommandPart.scala [hunk 2]
```diff
@@ -43,3 +44,29 @@ case class CwlArgumentCommandPart(argument: CommandLineBinding) extends CommandP
...
 }
 
+case class InputParameterCommandPart(argument: CommandInputParameter) extends CommandPart {
+
+  override def instantiate(inputsMap: Map[LocalName, WomValue],
+                           functions: IoFunctionSet,
+                           valueMapper: (WomValue) => WomValue) = {
+
+    val localInputsMap = inputsMap.map{
+      case (LocalName(localName), WomSingleFile(path)) => localName -> WomString(path)
```
Comment (cjllanwarne) [inline cwl/src/main/scala/cwl/CwlExpressionCommandPart.scala]: Just FMI, could you explain this?

File: cwl/src/main/scala/cwl/CwlExpressionCommandPart.scala [hunk 3]
```diff
@@ -43,3 +44,29 @@ case class CwlArgumentCommandPart(argument: CommandLineBinding) extends CommandP
   }
 }
 
+case class InputParameterCommandPart(argument: CommandInputParameter) extends CommandPart {
+
+  override def instantiate(inputsMap: Map[LocalName, WomValue],
+                           functions: IoFunctionSet,
+                           valueMapper: (WomValue) => WomValue) = {
+
+    val localInputsMap = inputsMap.map{
```
Comment (cjllanwarne) [inline cwl/src/main/scala/cwl/CwlExpressionCommandPart.scala]: ToL: Rather than preprocessing the entire list ahead of time, why not make a method (`eg "formatForCommandInsertion(womValue: WomValue)"`) and call it once we know which value it is we want to insert (on line 62)?

File: cwl/src/main/scala/cwl/CwlExpressionCommandPart.scala [hunk 4]
```diff
@@ -43,3 +44,29 @@ case class CwlArgumentCommandPart(argument: CommandLineBinding) extends CommandP
...
+      case (LocalName(localName), WomSingleFile(path)) => localName -> WomString(path)
+      case (LocalName(localName), value) => localName -> value
+    }
+
+
+    val womValue: WomValue = argument match {
+      case CommandInputParameter(id, _,_,_,_,_, Some(CommandLineBinding(None,Some(_),None,None,None,None,None)),_,_) =>
```
Comment (Horneth) [inline cwl/src/main/scala/cwl/CwlExpressionCommandPart.scala]: I wonder if at some point it just becomes better to just do `if blah.isDefined`, `blah.get` :smile:

Comment (danbills) [reply cwl/src/main/scala/cwl/CwlExpressionCommandPart.scala]: This is done because I have nested options, i.e. command line binding and the position of the command line binding.

I can't match the case where there is a command line binding *and* the command line binding has a position binding with an if statement but I can with a pattern match.

Comment (danbills) [reply cwl/src/main/scala/cwl/CwlExpressionCommandPart.scala]: The Nones are probably too strict.  I'm looking to see what I can make `_`

Comment (Horneth) [reply cwl/src/main/scala/cwl/CwlExpressionCommandPart.scala]: What about adding a method on `CommandInputParameter` like 
```scala
def hasPosition = inputBinding.exists(_.position.isDefined)
```

Comment (danbills) [reply cwl/src/main/scala/cwl/CwlExpressionCommandPart.scala]: So it turns out all those `None`'s being enforced is correct.  This case only handles when the input is bound by the id and not by any of the (many) options in `CommandLineBinding`.

Now I could write a function enforcing all this too, but it would only ever be used here and I would use a pattern match to implement that, too.  

I promise I would make this cleaner if I could :).

File: cwl/src/main/scala/cwl/CommandLineTool.scala [hunk 5]
```diff
@@ -101,6 +102,12 @@ case class CommandLineTool private(
   }
 
   def asCwl = Coproduct[Cwl](this)
+
+  def sortedInputs: Seq[CommandPart] =
```
[the remark this answers is not in the document]
Comment (danbills) [reply cwl/src/main/scala/cwl/CommandLineTool.scala]: Roger, I think `orderedInputs` is best.

File: cwl/src/main/scala/cwl/CommandLineTool.scala [hunk 6]
```diff
@@ -101,6 +102,12 @@ case class CommandLineTool private(
...
+    inputs.
+      sortBy(_.inputBinding.flatMap(_.position).getOrElse(0)).
```
[the remark this answers is not in the document]
Comment (danbills) [reply cwl/src/main/scala/cwl/CommandLineTool.scala]: Yep!  See `position` in http://www.commonwl.org/v1.0/CommandLineTool.html#CommandLineBinding

[the remark this answers is not in the document]
Comment (danbills) [reply cwl/src/main/scala/cwl/CommandLineTool.scala]: It now occurs to me that this is best immortalized in a comment.

**Comment by Horneth** [comment]
👍
